File: src/cspfj/filter/AC3.java
package cspfj.filter;

import java.util.BitSet;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

import cspfj.constraint.AbstractPVRConstraint;
import cspfj.constraint.Constraint;
import cspfj.heuristic.VariableHeuristic;
import cspfj.problem.Problem;
import cspfj.problem.Variable;

/**
 * @author scand1sk
 * 
 */
public final class AC3 implements Filter {
	private final Problem problem;

	private final BitSet inQueue;

	public int effectiveRevisions = 0;

	public int uselessRevisions = 0;

	private static final Logger logger = Logger.getLogger(Filter.class
			.getSimpleName());

	public AC3(final Problem problem, final VariableHeuristic heuristic) {
		super();
		this.problem = problem;

		inQueue = new BitSet(problem.getMaxVId() + 1);
	}

	public boolean reduceAll(final int level) {
		addAll();
		return reduce(level);

	}

	public boolean reduceAfter(final int level, final Variable variable) {
		if (variable == null) {
			return true;
		}
		for (int i = inQueue.nextSetBit(0); i >= 0; i = inQueue
				.nextSetBit(i + 1)) {
			final Variable v = problem.getVariable(i);
			final Constraint[] involved = v.getInvolvingConstraints();
			for (int j = involved.length; --j >= 0;) {
				involved[j].setRemovals(v.getPositionInConstraint(j), false);
			}
		}

		inQueue.clear();

		inQueue.set(variable.getId());

		final Constraint[] involving = variable.getInvolvingConstraints();

		for (int cp = involving.length; --cp >= 0;) {
			involving[cp].setRemovals(variable.getPositionInConstraint(cp),
					true);

		}

		return reduce(level);
	}

	private RevisionHandler revisator = new RevisionHandler() {
		public void revised(final Constraint constraint, final Variable variable) {
			inQueue.set(variable.getId());
			final Constraint[] involvingConstraints = variable
					.getInvolvingConstraints();

			for (int cp = involvingConstraints.length; --cp >= 0;) {
				final Constraint constraintP = involvingConstraints[cp];
				if (constraintP != constraint) {
					constraintP.setRemovals(variable
							.getPositionInConstraint(cp), true);
				}

			}
		}
	};

	private boolean reduce(final int level) {
		logger.finer("Reducing");
		final BitSet inQueue = this.inQueue;
		final RevisionHandler revisator = this.revisator;

		while (!inQueue.isEmpty()) {
			final Variable variable = pullVariable();

			final Constraint[] constraints = variable.getInvolvingConstraints();

			for (int c = constraints.length; --c >= 0;) {
				final Constraint constraint = constraints[c];

				final int position = variable.getPositionInConstraint(c);

//				if (!constraint.getRemovals(position)) {
//					constraint.fillRemovals(false);
//					continue;
//				}

				if (!constraint.revise(level, revisator)) {
					effectiveRevisions++;
					constraint.incWeight();
					constraint.fillRemovals(false);
					return false;
				}

				constraint.fillRemovals(false);

			}

		}

		assert control(level);

		return true;

	}

	private void addAll() {
		for (Variable v : problem.getVariables()) {
			inQueue.set(v.getId());
		}

		for (Constraint c : problem.getConstraints()) {
			c.fillRemovals(true);
		}
	}

	private boolean control(int level) {
		for (Constraint c : problem.getConstraints()) {
			if (!(c instanceof AbstractPVRConstraint)) {
				continue;
			}
			for (int i = c.getArity(); --i >= 0;) {
				if (!c.getVariable(i).isAssigned()
						&& ((AbstractPVRConstraint) c).revise(i, level)) {
					return false;
				}
			}
		}
		return true;
	}

	private Variable pullVariable() {
		Variable bestVariable = null;
		int bestValue = Integer.MAX_VALUE;

		final BitSet inQueue = this.inQueue;

		final Problem problem = this.problem;

		for (int i = inQueue.nextSetBit(0); i >= 0; i = inQueue
				.nextSetBit(i + 1)) {
			final Variable variable = problem.getVariable(i);
			final int domainSize = variable.getDomainSize();
			if (domainSize < bestValue) {
				bestVariable = variable;
				bestValue = domainSize;
			}
		}

		inQueue.clear(bestVariable.getId());
		return bestVariable;
	}

	public String toString() {
		return "GAC3rm";
	}

	public Map<String, Object> getStatistics() {
		final Map<String, Object> statistics = new HashMap<String, Object>(2);
		statistics.put("gac3-effective-revisions", effectiveRevisions);
		statistics.put("gac3-useless-revisions", uselessRevisions);
		return statistics;
	}

	@Override
	public boolean ensureAC() {
		return true;
	}
}

File: src/main/scala/cspfj/filter/AC3.scala
package cspfj.filter;

import java.util.Queue
import scala.annotation.tailrec
import cspfj.constraint.Constraint
import cspfj.constraint.Removals
import cspfj.priorityqueues.Key
import cspfj.priorityqueues.Fifos
import cspfj.priorityqueues.ScalaIOBinomialHeap
import cspfj.priorityqueues.ScalaNative
import cspfj.problem.Problem
import cspfj.problem.Variable
import cspfj.util.Loggable
import cspfj.Parameter
import cspfj.ParameterManager
import cspfj.Statistic
import cspfj.priorityqueues.BinomialHeap
import cspfj.priorityqueues.BinaryHeap
import cspfj.UNSATException

/**
 * @author scand1sk
 *
 */
object AC3 {
  @Parameter("ac3v.queue")
  var queueType: Class[_ <: Queue[Variable]] = classOf[BinaryHeap[Variable]]

  @Parameter("ac3v.key")
  val keyType: Class[_ <: Key[Variable]] = classOf[cspfj.heuristic.revision.Dom]

  def key = keyType.getConstructor().newInstance()

  def queue = queueType.getConstructor(classOf[Key[Variable]]).newInstance(key)

  ParameterManager.register(this);
}

final class AC3(
  val problem: Problem,
  val queue: Queue[Variable]) extends Filter with Loggable {

  // private static final Logger LOGGER = Logger.getLogger(Filter.class
  // .getSimpleName());

  @Statistic
  var revisions = 0;

  def this(problem: Problem) =
    this(problem, AC3.queue)

  def reduceAll() = {
    queue.clear();
    problem.variables.foreach(queue.offer)
    problem.constraints.foreach { c =>
      (0 until c.arity).foreach(c.setRemovals)
    }

    reduce()

  }

  private def prepareQueue(v: Variable) {
    queue.offer(v)
    v.constraints.zipWithIndex.foreach {
      case (c, i) =>
        c.setRemovals(v.positionInConstraint(i))
    }
  }

  private def setRemovals(v: Variable, skip: Constraint) {

    val constraints = v.constraints

    for (i <- 0 until constraints.size; val c = constraints(i) if c ne skip) {
      c.setRemovals(v.positionInConstraint(i))
    }

  }

  def updateQueue(prev: Array[Int], constraint: Constraint) {
    /** Requires high optimization */

    val scope = constraint.scope

    for (i <- prev.indices) {
      val variable = scope(i)
      if (prev(i) != variable.dom.size) {
        queue.offer(variable)
        setRemovals(variable, constraint)
      }
    }

  }

  @tailrec
  private def prepareQueue(modifiedConstraints: Iterator[Constraint]): Boolean = {
    if (modifiedConstraints.hasNext) {
      val c = modifiedConstraints.next
      (0 until c.arity).foreach(c.setRemovals)

      val prev = c.sizes()

      val sat = try {
        if (c.revise()) updateQueue(prev, c)

        c.clearRemovals()
        true
      } catch {
        case _: UNSATException => {
          c.weight += 1
          false
        }
      }

      sat && prepareQueue(modifiedConstraints)

    } else true

  }

  def reduceAfter(constraints: Iterable[Constraint]): Boolean = {
    Removals.clear()
    queue.clear()
    prepareQueue(constraints.iterator) && reduce()
  }

  def reduceAfter(variable: Variable) = variable == null || {
    Removals.clear()
    queue.clear()
    prepareQueue(variable)
    reduce()
  }

  def reduceFrom(modVar: Array[Int], modCons: Array[Int], cnt: Int): Boolean = {
    Removals.clear()
    queue.clear();
    // LOGGER.fine("reduce after " + cnt);

    for (v <- problem.variables if modVar(v.getId) > cnt)
      prepareQueue(v)

    if (modCons == null || prepareQueue(problem.constraints.iterator.filter(c => modCons(c.getId) > cnt))) {
      reduce()
    } else {
      false
    }

  }

  @tailrec
  private def reduce(): Boolean = {
    // LOGGER.finer("Reducing");
    if (queue.isEmpty) {
      assert(AC3Constraint.control(problem))
      true
    } else {
      val variable = queue.poll()
      //info(variable.toString)
      reduce(variable.constraints) && reduce()
    }

  }

  private def reduce(constraints: Array[Constraint]) = constraints.forall { c =>
    c.isEntailed || {
      revisions += 1
      val prev = c.sizes()
      //println("Revising " + c)

      try {
        if (c.revise()) {
          assert(!(c.sizes() sameElements prev), c + " returned wrong true revised info")
          updateQueue(prev, c)
        } else assert(c.sizes() sameElements prev, c + " returned wrong false revised info")

        c.clearRemovals();
        true
      } catch {
        case e: UNSATException =>
          c.weight += 1
          false
      }
    }
  }

  override def toString =
    "GAC3rm-var-" + queue.getClass.getSimpleName

  def getStatistics = Map("Revisions" -> revisions)

}
